--- mapnik-0.7.1/SConstruct.orig	2011-06-16 20:59:28.000000000 +0200
+++ mapnik-0.7.1/SConstruct	2011-06-16 21:02:18.816618308 +0200
@@ -1011,7 +1011,7 @@ if not preconfigured:
                 env.Append(CXXFLAGS = common_cxx_flags + '-O %s' % ndebug_flags)
         else:
             # Common flags for GCC.
-            gcc_cxx_flags = '-ansi -Wall %s -ftemplate-depth-100 %s' % (pthread, common_cxx_flags)
+            gcc_cxx_flags = '-ansi -Wall %s -ftemplate-depth-100 -DBOOST_FILESYSTEM_VERSION=2 %s' % (pthread, common_cxx_flags)
         
             if env['DEBUG']:
                 env.Append(CXXFLAGS = gcc_cxx_flags + '-O0 -fno-inline %s' % debug_flags)
