--- mapnik-v3.0.12/include/mapnik/text/harfbuzz_shaper.hpp.orig	2017-04-27 21:10:31.804876433 +0200
+++ mapnik-v3.0.12/include/mapnik/text/harfbuzz_shaper.hpp	2017-04-27 21:11:23.065634158 +0200
@@ -55,11 +55,7 @@
 static inline const uint16_t * uchar_to_utf16(const UChar* src)
 {
    static_assert(sizeof(UChar) == sizeof(uint16_t),"UChar is eq size to uint16_t");
-#if defined(_MSC_VER)
    return reinterpret_cast<const uint16_t *>(src);
-#else
-   return src;
-#endif
 }
 
 struct harfbuzz_shaper
