--- mapnik-3.0.9/SConstruct~	2015-11-26 11:13:12.000000000 +0100
+++ mapnik-3.0.9/SConstruct	2015-12-14 00:16:09.848114888 +0100
@@ -755,10 +755,6 @@
         # which goes <toolset>-<threading>-<abi>-<version>. See:
         #  http://www.boost.org/doc/libs/1_35_0/more/getting_started/unix-variants.html#library-naming
         append_params = ['']
-        if env['BOOST_TOOLKIT']: append_params.append(env['BOOST_TOOLKIT'])
-        if thread_flag: append_params.append(thread_flag)
-        if env['BOOST_ABI']: append_params.append(env['BOOST_ABI'])
-        if env['BOOST_VERSION']: append_params.append(env['BOOST_VERSION'])
 
         # Constructing the BOOST_APPEND setting that will be used to find the
         # Boost libraries.
